lib/regular_expression/ast/to_nfa.rb

```diff
@@ -0,0 +1,215 @@
+            when %q{\W}
+              start.add_transition(NFA::Transition::Invert.new(finish, [*("a".."z"), *("A".."Z"), *("0".."9"), "_"]))
+            when %q{\d}
+              start.add_transition(NFA::Transition::Range.new(finish, "0", "9"))
+            when %q{\D}
+              start.add_transition(NFA::Transition::Range.new(finish, "0", "9", invert: true))
+            when %q{\h}
+              start.add_transition(NFA::Transition::Range.new(finish, "a", "f"))
+              start.add_transition(NFA::Transition::Range.new(finish, "A", "F"))
+              start.add_transition(NFA::Transition::Range.new(finish, "0", "9"))
+            when %q{\H}
+              start.add_transition(NFA::Transition::Invert.new(finish, [*("a".."h"), *("A".."H"), *("0".."9")]))
+            when %q{\s}
+              start.add_transition(NFA::Transition::Value.new(finish, " "))
+              start.add_transition(NFA::Transition::Value.new(finish, "\t"))
+              start.add_transition(NFA::Transition::Value.new(finish, "\r"))
+              start.add_transition(NFA::Transition::Value.new(finish, "\n"))
+              start.add_transition(NFA::Transition::Value.new(finish, "\f"))
+              start.add_transition(NFA::Transition::Value.new(finish, "\v"))
+            when %q{\S}
+              start.add_transition(NFA::Transition::Invert.new(finish, [" ", "\t", "\r", "\n", "\f", "\v"]))
+            else
+              raise
+            end
+          when CharacterType
+            start.add_transition(NFA::Transition::Type.new(finish, node.value))
+          when Character
+            start.add_transition(NFA::Transition::Value.new(finish, node.value))
+          when Period
+            transition = NFA::Transition::Any.new(finish)
+            start.add_transition(transition)
+          when PositiveLookahead
+            start.add_transition(NFA::Transition::PositiveLookahead.new(finish, node.value))
+          when NegativeLookahead
+            start.add_transition(NFA::Transition::NegativeLookahead.new(finish, node.value))
+          when CharacterRange
+            transition = NFA::Transition::Range.new(finish, node.left, node.right)
+            start.add_transition(transition)
+          when Anchor
+            transition =
+              case node.value
+              when %q{\A}
+                NFA::Transition::BeginAnchor.new(finish)
+              when %q{\z}, %q{$}
+                NFA::Transition::EndAnchor.new(finish)
+              end
+
+            start.add_transition(transition)
+          when AddEpsilon
+            start.add_transition(NFA::Transition::Epsilon.new(finish))
+          else
+            raise "Unexpected work item: #{node.inspect}"
+          end
+        end
+
+        @start_state
+      end
+
+      private
+
+      def pop_work
+        @worklist.pop
+      end
+
+      def push_work(work)
+        work.reverse_each { |work_item| @worklist << work_item }
+      end
```

Comment:
We want to keep the exact traversal order, which was depth-first. Take this example:
```
     Root
  n1       n4
n2  n3   n5  n6
```

The order we want to process nodes in is in numerical order. Pushing *without* reversing would not result the desired order -- example:
- Process Root. Worklist = [n1, n4]
- Process n4, Worklist = [n1, n5, n6]
- ...

By reversing the items when pushing them to the worklist, we get the desired order:
- Process Root. Worklist = [n4, n1]
- Process n1. Workist = [n4, n3, n2]
- Process n2. Workist = [n4, n3]
- ...


We are changing the end of the array (push, pop) so that we don't have to shift every element when adding new work items, which would be costly.